### ISSUE TITLE
Implement portfolio gallery

### ISSUE THREAD

**Issue by ab-emptycup** [EmptyCupHQ/portfolio#14]
Every Professional's portfolio has a **Gallery**. Gallery is just a collection of images. Each image has an associated thumbnail image, a line of text description, upload timestamp.

| Method | Route | Operation |
| -------- | ----- | ---------- | 
GET |   /pros/__pid__/gallery/  |  Fetch images in the gallery |
POST |   /pros/__pid__/gallery/ | Upload an image to the gallery |
DELETE | /pros/__pid__/gallery/__imgid__/ | Remove an image from the gallery |

GET should return a JSON list of Image objects containing the keys:

- __id__: 16 char hex string (`_md5(uuid)[:16]` should do) identifying the image 
- __ref__: A https link where the image can be downloaded. Images should be stored separate in a Blob Store (S3 / ABS / GCS).
- __thumbnail_ref__: A https link where the thumbnail can be downloaded (Blob link)
- __description__: An optional textual description of the image (null, if not available)
- __uploaded_at__: Timestamp of when the image was uploaded to gallery

POST should upload the image file along with the image metadata.
DELETE should delete the image details from DB, remove the image and corresponding thumbnail from Blob Store

**Comment by Ashrjz** [comment]
@ab-emptycup
- For the JSON list of image objects containing data of each image related to a `pid`, will gallery be a field in the `pros` _db.json_, or gallery will have its JSON file having the same `uid` with the pro it is associated to?
- Currently for development, how will a blob store work?

**Comment by ab-emptycup** [comment]
>     * For the JSON list of image objects containing data of each image related to a `pid`, will gallery be a field in the `pros` _db.json_, or gallery will have its JSON file having the same `uid` with the pro it is associated to?

Yes, `gallery` will be a field in _db.json_. It will be a JSON encoded as string. When we move to SQL, the column will be of type "JSON".

>     * Currently for development, how will a blob store work?

Good question. A few examples of a blob store are: Amazon S3, Azure Blob Service Google Cloud Storage etc.
_store.py_ will be a module that we will write which will abstract the store functionality. For eg: 

- Uploading and saving a local file into the blob store.
- Checking if a given file exists in the blob container.
- Getting the downloadable URL of a file already present in the store.

_store.py_ module will give a generic interface that is independent of the underlying service provider (Amazon/Azure/GCP). To start with we'll implement a module _abs.py_ which will use Azure's python SDK to connect to storage container on Azure. _store.py_ will simply do `from abs import *`. That way, if tomorrow someone wants to deploy the system in AWS, they just have to change the `provider` option in `config.yml` and implement _s3.py_.

To get this to work, you may have to setup an azure account. If I remember correctly, Azure needs a debit card to sign up, even though nothing is charged. Hope that works for you... Do let me know if you want me to explain int better terms.

**Comment by Ashrjz** [comment]
@ab-emptycup 
- Are the thumbnail and the main image have different containers having the same storage name in azure?
- For image data, the input request is `multipart/form-data`, does that work?
- For the thumbnail what should be the resolution?
- And just a python util function for creating thumbnails works?
- What type of connection string to connect to the Azure storage?

**Comment by ab-emptycup** [comment]
On Sun, 19 Mar, 2023, 18:17 Ashish Raj, ***@***.***> wrote:

>
>    - Are the thumbnail and the main image have different containers
>    having the same storage name in azure?
>
> Either way works. If we keep different containers, we have to manage
containers. Else, we have to infer based on file name.

There might not be too much difference either way.


>    - For image data, the input request is multipart/form-data, does that
>    work?
>
> Yeah that should work I think.

>
>    - For the thumbnail what should be the resolution?
>
> That should be a config option. I think YouTube 144p is good enough.


>    - And just a python util function for creating thumbnails works?
>
> Yes, that should be fine. There are more complex and sophisticated ways to
support thumbnails, but this is probably the simplest and maybe even the
most efficient.


>    - What type of connection string to connect to the Azure storage?
>
> Environment Credential should be used. The required credentials have to be
set as environment variables while deploying.

**Comment by Ashrjz** [comment]
If we are dropping the `abs` module, the `config` module can store the token details using the credential to access Azure and the same for the container client objects, but where will the predefined methods of `azure.storage.blob` package, to upload or delete from the blob will be called because if they are called in the `store` module, the `store` module's independent dynamicity will be lost.
 
So, will all the upload or delete methods be in a class with the azure credentials be in a class in the `config` module?

Also, I think it would be better to have the `abs` module so that we can put everything in an azure class in `abs` and import it in the `config` module rather than importing it directly in the `store` module. This will maintain the generic interface and also reduce the dependency on the amount of code written in the `config` module.

**Comment by ab-emptycup** [comment]
On Fri, Apr 7, 2023 at 6:41 PM Ashish Raj ***@***.***> wrote:

> If we are dropping the abs module, the config module can store the token
> details using the credential to access Azure and the same for the container
> client objects, but where will the predefined methods of
> azure.storage.blob package, to upload or delete from the blob will be
> called because if they are called in the store module, the store module's
> independent dynamicity will be lost.
>
By independent dynamicity, I guess you mean the ability to support blob
stores from different clouds like GCP or AWS? If so, yes. It will not have
that abstraction. But, that is easy to separate out when we are adding
support for one of those clouds. If we separate "store" and "abs" now, it
will be over engineering as we don't have the requirement now. Infact, if
we separate out now, we may realise later that this separation is not clean
and we may have to refactor again after considering the api of the second
cloud. We might as well separate when we know what is needed.

So, let's have the azure functionality in "store" module itself for now.
When we add a second store, we'll separate both of them out.


> So, will all the upload or delete methods be in a class with the azure
> credentials be in a class in the config module?
>
Credentials and functionality should be separated. Credentials vary from
project to project & deployment to deployment. Functionality remains
constant. Functionality should be in store.py and credentials should be in
a flat dict object in config.py


> Also, I think it would be better to have the abs module as if we are
> putting everything in an azure class and import it in the config module
> rather than importing it directly in the store module. This will maintain
> the generic interface and also reduce the dependency on the amount of code
> written in the config module.
>
Refer above.
